internal/consensus/reactor.go

```diff
@@ -229,34 +229,43 @@ func (r *Reactor) OnStop() {
 		r.state.Wait()
 	}
 
-	r.mtx.Lock()
-	peers := r.peers
-	r.mtx.Unlock()
-
-	// wait for all spawned peer goroutines to gracefully exit
-	for _, ps := range peers {
-		ps.closer.Close()
-	}
-	for _, ps := range peers {
-		ps.broadcastWG.Wait()
-	}
+	// Close closeCh to signal to all spawned goroutines to gracefully exit. All
+	// p2p Channels should execute Close().
+	close(r.closeCh)
 
 	// Close the StateChannel goroutine separately since it uses its own channel
 	// to signal closure.
 	close(r.stateCloseCh)
 	<-r.stateCh.Done()
 
-	// Close closeCh to signal to all spawned goroutines to gracefully exit. All
-	// p2p Channels should execute Close().
-	close(r.closeCh)
-
 	// Wait for all p2p Channels to be closed before returning. This ensures we
 	// can easily reason about synchronization of all p2p Channels and ensure no
 	// panics will occur.
 	<-r.voteSetBitsCh.Done()
 	<-r.dataCh.Done()
 	<-r.voteCh.Done()
 	<-r.peerUpdates.Done()
+
+	peers := make(map[types.NodeID]*PeerState)
+	r.mtx.Lock()
+
+	// Here, we make a copy of the map of peer states.
+	// The code below proceeds to access entries from the map outside of a lock.
+	// Goroutines launched by other functions may be deleting from the map but should
+	// no longer be appending to it since we waited for the peerUpdates channel to complete.
+	// Each of the methods below is safe to perform concurrently on the peers.
+	for id, state := range r.peers {
+		peers[id] = state
+	}
+	r.mtx.Unlock()
+
+	// wait for all spawned peer goroutines to gracefully exit
```

Comment:
Could we do the closes inside the above loop, while we already hold the lock?

Reply:
Also, I've been assuming that the Wait calls can't complete without holding the lock, but if that's not true, we might be able to avoid needing to copy at all—do the peers update their own map entry or do they just hold the state object?

Reply:
Each `PeerState` has its own lock for managing the fields on `PeerState`. They do not update their own entry in the map. The routines that manage the `WaitGroup` also do not take the lock: https://github.com/tendermint/tendermint/blob/00e8c0e09a86b8c523386010ae4b0d874121a977/internal/consensus/reactor.go#L977

It therefore appears safe to wait in the lock.